### DISCUSSION TITLE
[Bug] skywalking-go sdk panic

### DISCUSSION THREAD

**Discussion by Donghui0** [apache/skywalking#11933] [Q&A]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Go Agent (apache/skywalking-go)

### What happened

Sporadic panic problem:
 panic: runtime error: invalid memory address or nil pointer dereference
 [signal SIGSEGV: segmentation violation code=0x1 addr=0x1d pc=0x884742]

 goroutine 68 [running]:
 google.golang.org/protobuf/internal/impl.sizeMessageSliceInfo({0x269?}, 0xc00e220748, {0xf0?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/codec_field.go:472 +0x42
 google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow(0xc000366e58, {0x20000000000001e?}, {0x3?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/encode.go:79 +0xd9
 google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer(0xc00a76adc0?, {0xc00cb22320?}, {0x58?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/encode.go:56 +0x76
 google.golang.org/protobuf/internal/impl.sizeMessageSliceInfo({0x200000000419dcb?}, 0xc00151ee60, {0x84?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/codec_field.go:473 +0x55
 google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow(0xc000366d10, {0xc000046077?}, {0x76?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/encode.go:79 +0xd9
 google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer(0x284?, {0xd?}, {0x80?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/encode.go:56 +0x76
 google.golang.org/protobuf/internal/impl.sizeMessageSliceInfo({0x7efedb8ab1b8?}, 0xc000295990, {0x68?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/codec_field.go:473 +0x55
 google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow(0xc000366a80, {0xc009e61a68?}, {0x80?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/encode.go:79 +0xd9
 google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer(0xc009e61ad8?, {0x9?}, {0x80?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/encode.go:56 +0x76
 google.golang.org/protobuf/internal/impl.(*MessageInfo).size(0x40802d?, {{}, {0x286dcb8?, 0xc00bbf7950?}, 0x5?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/internal/impl/encode.go:40 +0x52
 google.golang.org/protobuf/proto.MarshalOptions.marshal({{}, 0x50?, 0x0, 0x0}, {0x0, 0x0, 0x0}, {0x286dcb8, 0xc00bbf7950})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/proto/encode.go:156 +0x142
 google.golang.org/protobuf/proto.MarshalOptions.MarshalAppend({{}, 0x60?, 0x63?, 0x1e?}, {0x0, 0x0, 0x0}, {0x285a880?, 0xc00bbf7950?})
 	/data/repo_cache/golang/mod/google.golang.org/protobuf@v1.30.0/proto/encode.go:125 +0x79
 github.com/golang/protobuf/proto.marshalAppend({0x0, 0x0, 0x0}, {0x7efeb03f7900?, 0xc00bbf7950?}, 0x80?)
 	/data/repo_cache/golang/mod/github.com/golang/protobuf@v1.5.3/proto/wire.go:40 +0xa5
 github.com/golang/protobuf/proto.Marshal(...)
 	/data/repo_cache/golang/mod/github.com/golang/protobuf@v1.5.3/proto/wire.go:23
 google.golang.org/grpc/encoding/proto.codec.Marshal({}, {0x21e6360, 0xc00bbf7950})
 	/data/repo_cache/golang/mod/google.golang.org/grpc@v1.55.0/encoding/proto/proto.go:45 +0x4e
 google.golang.org/grpc.encode({0x7efedb8b28f8?, 0x3952e88?}, {0x21e6360?, 0xc00bbf7950?})
 	/data/repo_cache/golang/mod/google.golang.org/grpc@v1.55.0/rpc_util.go:632 +0x44
 google.golang.org/grpc.prepareMsg({0x21e6360?, 0xc00bbf7950?}, {0x7efedb8b28f8?, 0x3952e88?}, {0x0, 0x0}, {0x0, 0x0})
 	/data/repo_cache/golang/mod/google.golang.org/grpc@v1.55.0/stream.go:1754 +0xd2
 google.golang.org/grpc.(*clientStream).SendMsg(0xc00ca3e900, {0x21e6360, 0xc00bbf7950})
 	/data/repo_cache/golang/mod/google.golang.org/grpc@v1.55.0/stream.go:875 +0x1b5
 skywalking.apache.org/repo/goapi/collect/language/agent/v3.(*traceSegmentReportServiceCollectClient).Send(0x2863028?, 0x2863098?)
 	/data/repo_cache/golang/mod/skywalking.apache.org/repo/goapi@v0.0.0-20230314034821-0c5a44bb767a/collect/language/agent/v3/Tracing_grpc.pb.go:61 +0x2b
 github.com/apache/skywalking-go/agent/reporter.(*gRPCReporter).initSendPipeline.func1()
 	grpc.go:323 +0x1eb
 created by github.com/apache/skywalking-go/agent/reporter.(*gRPCReporter).initSendPipeline
 	grpc.go:305 +0x6a

### What you expected to happen

no panic. Normal operation, no impact on business.

### How to reproduce

Occasional problems.

### Anything else

go ver: 1.18.10

go.mod:
	google.golang.org/grpc v1.55.0 // indirect
        google.golang.org/protobuf v1.30.0 // indirect

### Are you willing to submit a pull request to fix on your own?

- [ ] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
I am not sure this stack is enough to tell what is the issue.

**Comment by mrproliu** [comment]
Could you help to adding a demo in there? It looks like a gRPC/protobuf error.

**Comment by Donghui0** [comment]
> Could you help to adding a demo in there? It looks like a gRPC/protobuf error.

GRPC Server is an Opentelmetry skywalking receiver.
opentelementry:  0.90

**Comment by wu-sheng** [comment]
Why to open telemetry rather than OAP? 

https://github.com/apache/skywalking/discussions/11916

That collector is not recommended to use.

**Comment by Donghui0** [comment]
> Why to open telemetry rather than OAP?
> 
> #11916
> 
> That collector is not recommended to use.

This component is not used. The overall architecture is based on opentelmetry. It looks like:
![image](https://github.com/apache/skywalking/assets/7134928/934438ba-2dae-41d8-8ef9-fd97084db6ba)

**Comment by wu-sheng** [comment]
Then why asking us? We don't pormise this works.

**Comment by wu-sheng** [comment]
Convert to the discussion as it is not the official way to work. 
Reach Alicloud as they suggest you to do so.

**Reply by Donghui0** [reply]
The problem occurs in the first half of the link, when go-sdk calls opentelmetry(skywalking reciver).
It has nothing to do with Alibaba Cloud.  The stacktrace shows errors of golang SDK, not opentelmetry.

**Reply by wu-sheng** [reply]
We don't promise about skywalking-go agent to OTEL collector, because we don't test that. And for our use cases, you need to ship the data to OAP side.

**Reply by wu-sheng** [reply]
We don't know how OTEL receiver would work with skywalking-go agent, we even don't know there is a receiver. 

That thing was added on Sep 3, 2022, https://github.com/open-telemetry/opentelemetry-collector-contrib/commits/5133f4ccd69fa40d016c5b7f2198fb6ac61007b4/receiver/skywalkingreceiver/README.md. We don't know that, and that belongs to OTEL.

I could only share that, skywalking-go was created in April 2023, https://github.com/apache/skywalking-go/graphs/contributors.

Reach that receiver maintainer, or your service vendor to get that out.
